### PR TITLE
Updates for the `preview-3` release

### LINKED ISSUE #341: Document preview-to-preview upgrade steps
When upgrading a .NET Aspire app from a previous preview, there are a few steps that need to happen:

1. Install the new workload version
2. Update all Aspire PackageReferences (especially the `Aspire.Hosting` reference in the AppHost.csproj) to match the version of the workload you are updating to
3. Update any code that used APIs that were removed/changed to the new APIs

We should have an upgrade section in our docs that list these steps, and have a trouble shooting section for common mistakes, and what errors you see.

One example of having a mis-matched workload and `Aspire.Hosting` PackageReference version is: https://github.com/dotnet/aspire/issues/1944

cc @IEvangelist

### LINKED ISSUE #357: For Aspire preview.3, we should update the setup instructions for using VS with aspire
For the aspire Setup and tooling documentation related to VS (i.e. [this document](https://learn.microsoft.com/en-us/dotnet/aspire/setup-tooling?tabs=visual-studio), we should remove the part where we ask users to run `dotnet workload update` and `dotnet workload install aspire`. These 2 commands will become redundant because VS 17.10 Preview 1 includes Aspire 8.0 P3. Instead, we can call out that VS 17.10 Preview 1 will include the latest Aspire workload and then have them simply run `dotnet workload list` after VS has been updated to check the version of Aspire that they have installed. 

cc: @IEvangelist 

---
[Associated WorkItem - 209382](https://dev.azure.com/msft-skilling/Content/_workitems/edit/209382)

### LINKED ISSUE #358: The What's new in .NET Aspire needs updating
We'd like to pivot a bit:

- The generated bits are fine, and can remain
- We should however add, product release notes, taking the place of both GitHub and .NET devblogs. 😬 

---
[Associated WorkItem - 209383](https://dev.azure.com/msft-skilling/Content/_workitems/edit/209383)

### PR DESCRIPTION
**Pull request by IEvangelist**
Addresses various issues related to the release of `preview-3`.

- Fixes #341
- Fixes #358
- Fixes #357

What's new in .NET Aspire Hub: https://review.learn.microsoft.com/en-us/dotnet/aspire/whats-new/?branch=pr-en-us-359

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/setup-tooling.md](https://github.com/dotnet/docs-aspire/blob/23ccd60314f650cb0813836d2efbb8db87fa7984/docs/fundamentals/setup-tooling.md) | [.NET Aspire setup and tooling](https://review.learn.microsoft.com/en-us/dotnet/aspire/fundamentals/setup-tooling?branch=pr-en-us-359) |
| [docs/whats-new/preview-3.md](https://github.com/dotnet/docs-aspire/blob/23ccd60314f650cb0813836d2efbb8db87fa7984/docs/whats-new/preview-3.md) | [.NET Aspire preview 3](https://review.learn.microsoft.com/en-us/dotnet/aspire/whats-new/preview-3?branch=pr-en-us-359) |

### CODE REVIEW

File: docs/whats-new/preview-3.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    *Note that if you had previous preview versions of the .NET Aspire workload already installed, you may see messages informing you that the workload is already installed. These can be safely ignored.*
+
+1. Refer to the docs for [detailed instructions on installing .NET Aspire](../fundamentals/setup-tooling.md)
+
+After updating you can run `dotnet workload list` to see the updated version (note: your version of Visual Studio may differ):
+
+```text
+Installed Workload Id    Manifest Version                   Installation Source
+-----------------------------------------------------------------------------------------
+aspire                   8.0.0-preview.3.XXXX.YY/8.0.100    SDK 8.0.100, VS 17.10.XXXX.YY
```
Comment (joperezr) [inline docs/whats-new/preview-3.md:38]:
```suggestion
aspire                   8.0.0-preview.3.24105.21/8.0.100    SDK 8.0.100, VS 17.10.XXXX.YY
```

File: docs/whats-new/preview-3.md [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+```
+
+### Update existing apps
+
+For existing .NET Aspire apps, after performing the above steps to get the latest workload, you'll want to change your package references for any Aspire components. Update **all Aspire package references (hosting and components)** to `8.0.0-preview.3.XXXXX.YY`, for example the hosting would change to:
```
Comment (joperezr) [inline docs/whats-new/preview-3.md:43]:
```suggestion
For existing .NET Aspire apps, after performing the above steps to get the latest workload, you'll want to change your package references for any Aspire components. Update **all Aspire package references (hosting and components)** to `8.0.0-preview.3.24105.21`, for example the hosting would change to:
```

File: docs/whats-new/preview-3.md [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+```xml
+<PackageReference Include="Aspire.Hosting" Version="8.0.0-preview.3.XXXX.YY" />
```
Comment (joperezr) [inline docs/whats-new/preview-3.md:46]:
```suggestion
<PackageReference Include="Aspire.Hosting" Version="8.0.0-preview.3.24105.21" />
```

File: docs/whats-new/preview-3.md [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+builder.Build().Run();
+```
+
+### `WithEntrypoint` is now supported for containers
+
+Using `WithEntrypoint` enables the entrypoint to be specified for container based resources.
+
+### RedisCommander resource extension
+
+[Redis commander](https://github.com/joeferner/redis-commander) is a web management tool for viewing the contents of the Redis Cache. With this change you can use the `WithRedisCommander` extension to both IResourceBuilder<RedisResource> and IResourceBuilder<RedisContainerResource> which results in a container resource being added to the model (once) and a lifecycle hook which configures it for Redis Commander.
```
Comment (IEvangelist) [inline docs/whats-new/preview-3.md:289]:
```suggestion
[Redis commander](https://github.com/joeferner/redis-commander) is a web management tool for viewing the contents of the Redis Cache. With this change you can use the `WithRedisCommander` extension to both `IResourceBuilder<RedisResource>` and `IResourceBuilder<RedisContainerResource>` which results in a container resource being added to the model (once) and a lifecycle hook which configures it for Redis Commander.
```

**Review by samsp-msft — APPROVED** [review]

File: docs/fundamentals/setup-tooling.md [hunk 5]
```diff
@@ -30,6 +30,14 @@ The .NET Aspire workload installs internal dependencies and makes available othe
 
 # [Visual Studio](#tab/visual-studio)
 
+Starting with Visual Studio 2022 17.10 Preview 1 includes the **.NET Aspire preview 3** workload by default. To verify that you have the .NET Aspire workload installed, run the following command:
```
Comment (alexwolfmsft) [inline docs/fundamentals/setup-tooling.md:33]:
```suggestion
Visual Studio 2022 17.10 Preview 1 or higher includes the **.NET Aspire preview 3** workload by default. To verify that you have the .NET Aspire workload installed, run the following command:
```

File: docs/whats-new/preview-3.md [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+Many small changes across the dashboard have been made to support localization and accessibility. These changes include improving color contrast, adding accessibility attributes, and moving UI text into resource files.
+
+An example of an improvement is how the dashboard indicates trace length. Previously there was a subtle color gradient on each row that didn't meet accessibility requirements. In preview 3 this was replaced with a radial progress icon:
+
+:::image type="content" source="media/preview-3/trace-duration.png" lightbox="media/preview-3/trace-duration.png" alt-text="Trace details page with new radial icon used to quickly differentiate trace lengths.":::
+
+### New resource details view
+
+The resources page grid was bursting at the seams with information. In preview 3 there's now a resource details view. The grid continues to display important information, and additional details, including all of a resources properties, are available in the details view.
```
Comment (alexwolfmsft) [inline docs/whats-new/preview-3.md:82]:
```suggestion
The resources page grid was previously overloaded with information. Preview 3 now includes a resource details view for improved readability. The grid continues to display important information, while the details view shows additional data, including all of the resource's properties.
```

Just a personal opinion but this bit sounded a little odd to me - an alternate version here.

File: docs/whats-new/preview-3.md [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+
+1. If on Windows and using Visual Studio to work with .NET Aspire, install/update to the latest [Visual Studio 2022 Preview](https://aka.ms/vspreview) release (17.10.0 Preview 1.0 at the time of writing).
+
+    - Ensure that the **.NET Aspire SDK (Preview)** component is selected under the **Individual components** tab of the Visual Studio installer:
+
+    ![Visual Studio installer with .NET Aspire SDK (Preview) component selected](media/preview-3/vs-aspire-sdk-component.png)
+
+1. If on Windows and you have Visual Studio 2022 17.8.x installed but intend to only use .NET Aspire via the .NET CLI (`dotnet`), download and install the [.NET 8.0.100 SDK using the standalone installer](https://dotnet.microsoft.com/download/dotnet/8.0).
+
+1. If on macOS or Linux, download and install the [.NET 8.0.100 SDK](https://dotnet.microsoft.com/download/dotnet/8.0).
```
Comment (alexwolfmsft) [inline docs/whats-new/preview-3.md:23]: Maybe not important for a looser what's new page, but 1-3 might read better as a set of tabs for Visual Studio, CLI, and macOS/Linux with the workload instructions below it.

File: docs/whats-new/preview-3.md [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+
+The dashboard telemetry pages show off a number of improvements:
+
+- **Metrics filter UI**: The metrics filter UI has been redesigned. It's now much faster to update filters when viewing a metrics chart.
+- **Histogram metrics count**: There is a new option to view the count of items recorded to a histogram metric.
+- **Span events**: OpenTelemetry spans support events. For example, when was the first data received during a database call. Events are available in span details view.
+
+## Component updates
+
+[.NET Aspire components](../fundamentals/components-overview.md) are a curated suite of NuGet packages specifically selected to facilitate the integration of cloud-native applications with prominent services and platforms. Each component furnishes essential cloud-native functionalities through either automatic provisioning or standardized configuration patterns. .NET Aspire components can be used without an orchestrator project, but they're designed to work best with the .NET Aspire app host.
```
Comment (alexwolfmsft) [inline docs/whats-new/preview-3.md:96]: Love this concise but informative summary of components.

File: docs/whats-new/preview-3.md [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+    return response.Value.Choices.FirstOrDefault()?.Message.Content
+        ?? "no response...";
+});
+
+app.MapDefaultEndpoints();
+
+app.Run();
+```
+
+For more information, see [GitHub: .NET Aspire.Azure.AI.OpenAI](https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.Azure.AI.OpenAI).
```
Comment (alexwolfmsft) [inline docs/whats-new/preview-3.md:131]: Does OpenAI need anything for the Hosting project?

File: docs/whats-new/preview-3.md (hunk 7 above)
Comment (IEvangelist) [reply docs/whats-new/preview-3.md:23]: I thought about that and had it that way originally. It felt a bit redundant with the changes to setup and tooling, which we link to - that have been updated. This was adapted from the original draft blog post, and I went with simplicity, and it felt like tabs was overkill.

File: docs/whats-new/preview-3.md (hunk 9 above)
Comment (IEvangelist) [reply docs/whats-new/preview-3.md:131]: Yeah, see https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting.Azure/AzureOpenAIResource.cs

File: docs/whats-new/preview-3.md [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+Once added, when run a pre-configured Redis Commander instance is spun up to interactively browse what is in the Redis Cache. It can be accessed via the Aspire Dashboard. For example:
+
+```csharp
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddRedis("myredis1").WithRedisCommander();
+
+builder.AddRedis("myredis2").WithRedisCommander();
+```
```
Comment (alexwolfmsft) [inline docs/whats-new/preview-3.md:321]: This is really coool.

File: docs/whats-new/preview-3.md [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+       .WithDaprSidecar(new DaprSidecarOptions
+        {
+            AppId = "serviceb",
+            RuntimePath = "/some/other/runtime/folder"
+        });
+```
+
+## Add support for Orleans in .NET Aspire
+
+Starting with preview 3, support for [Orleans](/dotnet/orleans/overview) was added, allowing you to configure a cluster in your app host, specifying the resources the cluster uses. For example, you can specify that an Azure Table will be used for cluster membership, an Azure Redis resource will be used for the grain directory, and an Azure Blob Storage resource will be used to store grain state. The integration currently support Redis and Azure Table & Blob storage resources. Support for other resources will be added later.
```
Comment (ReubenBond) [inline docs/whats-new/preview-3.md:367]:
```suggestion
Preview 3 adds support for [Orleans](/dotnet/orleans/overview) v8.1.0-preview1 or later, allowing you to configure a cluster in your app host, specifying the resources the cluster uses. For example, you can specify that an Azure Table will be used for cluster membership, an Azure Redis resource will be used for the grain directory, and an Azure Blob Storage resource will be used to store grain state. The integration currently support Redis and Azure Table & Blob storage resources. Support for other resources will be added later.
```

**Pull request merged**